client/active-game/active-game-manager.js

```diff
@@ -50,15 +51,28 @@ export default class ActiveGameManager extends EventEmitter {
         // Same config as before, no operation necessary
         return current.id
       }
-      // Quit the currently active game so we can replace it
-      this.emit('gameCommand', current.id, 'quit')
+      if (current.id !== config.gameId) {
```

Comment:
I wonder if this should come before the `deepEqual` check (and that should be an `else if`)? Unsure of what the semantics are here as far as whether a different config would always mean a different ID or whatever. I also don't remember where these IDs get generated now, but if it's a server thing, maybe the config check is completely unnecessary (because the server would handle config changes by making a new ID).

Reply:
I assume this used to handle the case where a game process would leave hanging for whatever reason, to give it a chance to exit the game one last time? We cleanup the active game (by setting it to `null`) both when a game exits cleanly and when there's a launch error, so not sure if it's worth having this check anymore, but I don't think it's hurting anything.

As for the `deepEqual` check, I don't think it's necessary anymore. The server is the one that generates the `gameId` now, and it will do so every time some service calls the `loadGame` function:
https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/games/game-loader.js#L83

The only time when the `deepEqual` check would return `true` is if the `setGameConfig` function is called twice during the loading of a same game? I don't see at the moment how that would be possible.

Reply:
Well before, the ID was generated on the client (and was a way of sending messages to specific clients). You can kind of think of it as a hash of the game config before though (as if the game config was the same, we'd simply keep using the ID we had). So yeah, I think now that that's gone, no `deepEqual` check is needed, you can simply check the ID for the same purpose

Reply:
One thing that crossed my mind by having this check if the `gameId` is different (and exiting the "old" game if it is), is that the users might have some unintended consequences. Since we currently don't have any checks that stops a user from starting a game while they're already in another game (and not sure that we should?), it's possible for a user to, for example, click on a replay while they're in another game and their game will abruptly exit and be replaced with a replay.

So I would do either of the following two things:
- remove this check that sends a `quit` command to the old game if a user is already in a game
- don't unsubscribe users from gameplay activity registry until _after_ the game finishes, not just until they leave the lobby/queue

I'm personally leaning towards the former, but I'm not sure if there are any unintended consequences of allowing users to be in multiple games at once.

Reply:
I'd prefer something like the latter, although I'm not actually sure if what you mentioned is the right approach. The code we have here has always assumed we only had one active game, and any additional active games were just things we needed to shut down. I don't see why we'd want to let players be in two multiplayer games at once, and watching a replay while playing a ranked game doesn't seem like something we'd want to support either (not to mention that with SC:R, the resource usage of doing that is quite high)

Reply:
Yeah, thinking about it a bit further, I think I agree with the latter option as well now.

I'll leave the check which exits the old game then for now, and eventually we'll probably need to extend the game loader to deal with registering/unregistering the gameplay activities, as well as handling other malfunctions and stuff as you mentioned.